src/hardware/sblaster.cpp

```diff
@@ -1013,19 +1013,65 @@ static std::vector<AudioFrame>& maybe_silence(const T* samples,
 
 static uint32_t read_dma_8bit(const uint32_t bytes_to_read, const uint32_t i = 0)
 {
-	const auto bytes_read = sb.dma.chan->Read(bytes_to_read, sb.dma.buf.b8 + i);
-	assert(bytes_read <= DmaBufSize * sizeof(sb.dma.buf.b8[0]));
+	static_assert(sizeof(sb.dma.buf.b8[0]) == 1);
+
+	if (i >= DmaBufSize) {
+		// Should never happen as the code is current written.
+		// Calling code has i either 0 or 1 to handle a dangling sample from the last read.
+		assertm(false, "SBLASTER: Read requested out of bounds of sb.dma.buf.b8");
+		return 0;
+	}
+
+	// Assert that the DMA controller is configured for 8 bit reads.
+	// so we don't have to do the whole song and dance we do for 16 bit.
+	// We should always be dealing with bytes here.
+	assertm(!sb.dma.chan->is_16bit, "SBLASTER: read_dma_8bit() called but DMA controller is configured for 8 bit reads");
+
+	const uint32_t bytes_available = DmaBufSize - i;
+	const uint32_t clamped_bytes = std::min(bytes_to_read, bytes_available);
+	const auto bytes_read = sb.dma.chan->Read(clamped_bytes, sb.dma.buf.b8 + i);
```

Comment:
C int promotion... oh man 🥶

Reply:
They're all `uint32_t`.  I don't think there's any promotion happening here?

Reply:
Yeah, I think you're right. If one of the operands is unsigned, then unsigned "wins" (for mixed signed/unsigned arithmetic). But this is pure unsigned, so that doesn't even come into the picture.

This is a good description of the main algo:

https://learn.microsoft.com/en-us/cpp/c-language/usual-arithmetic-conversions?view=msvc-170